Add support for topics with multiple schemas (#175)

mkubala opened the pull request:
This contribution makes kafka-connect-bigquery compatible with topics that carry different types of messages, especially when Kafka is used for an event sourcing.

This approach is already supported by Kafka and Schema Registry (see https://www.confluent.io/blog/put-several-event-types-kafka-topic/).

What's been changed:
* Added support for topics with multiple record types (schemas) by adding support for TopicAndRecordName strategy,
* Added support for creating missing tables on the fly (no need to restart connector whenever new record type is being published to the multi schema topic),
* Moved `autoCreateTables` setting from `BigQuerySinkConnectorConfig` to `BigQuerySinkConfig`,
* After that `BigQuerySinkConnectorConfig` become no longer needed so I removed it,
* Added support for handling empty records (for both, SinkRecord and Schema converters) - for event envelopes with an empty payload.
* Cleaned code from unused methods and fields in various places

C0urante commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConnectorConfig.java:
```diff
@@ -1,82 +0,0 @@
-/**
- * Class for connector-specific configuration properties.
- */
-public class BigQuerySinkConnectorConfig extends BigQuerySinkConfig {
```
Good night, sweet prince 😞

mkubala commented:
I'm going to add yet another improvement - when connector is running in a multi schema mode, we should be able to match dataset based on the topic and record name. This way we will be able to aggregate tables by their logical meaning (e.g. payment_events dataset with tables for successful and failed payments, order_events containing tables for orders' state changes etc.).

I decided to do it in a separate PR because this one is in my opinion already too large (sorry for that), but it would be great if you could include both of them in one release (or make subsequent release right after another PR will be approved and merged, which is what I really hope for 😉 ).

mkubala commented:
I forgot that I'm working on a forked repo and my feature branch is absent here, so I'm not able to expose a PR that's based on top of this particular branch. Looks like I have to add another commit to this PR.

jaroslawZawila commented:
What is going on with this PR?

mkubala commented:
I see that in the meantime a couple of other PRs have been merged and because this PR is quite big, conflicts appear. I'm going to fix them, hoping this PR will be reviewed soon :) cc  @mtagle @criccomini @C0urante

criccomini commented:
Hey all, just saw this. Sorry, lemme follow up with @bingqinzhou

criccomini commented:
This is a pretty big PR. @mkubala can you break it up?

bingqinzhou commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:
```diff
@@ -131,11 +132,23 @@ private PartitionedTableId getRecordTable(SinkRecord record) {
+      TopicToTableResolver.updateTopicToTable(config, topicAndRecordName, topicsToBaseTableIds);
+      TableId tableId = topicsToBaseTableIds.get(topicAndRecordName);
+      if (bigQuery.getTable(tableId) == null) {
+        boolean createMissingTables = config.getBoolean(BigQuerySinkConfig.TABLE_CREATE_CONFIG);
```
Can you please change the boolean name to be `autoCreateTables` for consistency?

mkubala replied:
sure

bingqinzhou commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:
```diff
@@ -131,11 +132,23 @@ private PartitionedTableId getRecordTable(SinkRecord record) {
+          logger.info("Table {} does not exist. Attempting to create.", tableId.getTable());
+          schemaManager.createTable(tableId, topicAndRecordName);
+        } else {
+          throw new ConnectException(
```
Do we need to specifically throw an exception here? Since if the table doesn't exist, bigquery will throw a corresponding exception while trying to insert records to that table.

bingqinzhou commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java:
```diff
@@ -608,6 +706,23 @@ private void verifyBucketSpecified() throws ConfigException {
+      );
+    }
+
+    if (schemaRetriever == null) {
```
Can you please remove this check since it's not necessary and confusing?

mtagle commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:
```diff
@@ -131,11 +132,23 @@ private PartitionedTableId getRecordTable(SinkRecord record) {
+        boolean createMissingTables = config.getBoolean(BigQuerySinkConfig.TABLE_CREATE_CONFIG);
+        if (createMissingTables) {
+          logger.info("Table {} does not exist. Attempting to create.", tableId.getTable());
+          schemaManager.createTable(tableId, topicAndRecordName);
```
this feels out of scope for this PR. Is there a pr-relevant reason you can't just create the necessary tables on start up like the connector does at the moment?

mkubala replied:
Yes, there is a reason, described in the previous comment (https://github.com/wepay/kafka-connect-bigquery/pull/219#discussion_r363647317)

mtagle commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java:
```diff
@@ -100,12 +110,29 @@
   private static final ConfigDef.Type TOPICS_TO_TABLES_TYPE =              ConfigDef.Type.LIST;
   private static final ConfigDef.Importance TOPICS_TO_TABLES_IMPORTANCE =
       ConfigDef.Importance.MEDIUM;
-  public static final Object TOPICS_TO_TABLES_DEFAULT =                    null;
```
these sorts of changes feel unnecessary and not related to this PR (and, conceptually, I think having the default for a particular config be public is fine).

mtagle commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConnectorConfig.java:
```diff
@@ -1,82 +0,0 @@
-/**
- * Class for connector-specific configuration properties.
- */
-public class BigQuerySinkConnectorConfig extends BigQuerySinkConfig {
```
What exactly is the reason for removing this?

mkubala replied:
The only member of this config (auto create tables) has been moved to its parent (`BigQuerySinkConfig`) because I had to access this property also from the task level.
After that `BigQuerySinkConnectorConfig` became empty.

mtagle commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/convert/BigQueryRecordConverter.java:
```diff
@@ -178,12 +178,17 @@ private Object convertObject(Object kafkaConnectObject, Schema kafkaConnectSchem
-      if (bigQueryObject != null) {
-        bigQueryRecord.put(kafkaConnectField.name(), bigQueryObject);
+      // ignore empty structures
+      boolean isEmptyStruct = kafkaConnectField.schema().type() == Schema.Type.STRUCT &&
```
I'm a little confused about what's going on here or how it relates to the PR.

mkubala replied:
Sometimes particular record types have fields with instances of classes with no fields (null objects, markers etc.).
Record and schema converters should omit such nested empty structs, otherwise BigQuery will raise an error.

criccomini commented:
Echoing sentiment here:

1. PR is too big. Please split it up.
2. We will merge @bingqinzhou's PR. Her approach is roughly similar to yours. Please rebase with her changes once her PR is merged in (planning tomorrow AM).

Sorry for delay on this. Thanks for contributing--it's appreciated. :)

mkubala commented on kcbq-api/src/main/java/com/wepay/kafka/connect/bigquery/api/SchemaRetriever.java:
```diff
@@ -16,22 +18,24 @@
    * Retrieve the most current schema for the given topic.
    * @param table The table that will be created.
-   * @param topic The topic to retrieve a schema for.
+   * @param topicAndRecordName The topic and an optional record name to retrieve a schema for.
```
Record name is optional. I did not want to end up with methods that have too many arguments and overloads. 

It's been a while since I commercially worked with Java (I use Scala on my daily basis) and AFAICS passing parameters of type `Optional<T>` is not recommended (as one of the Java core team members, it's not aimed to be a counterpart of `Maybe` monad from Haskell or Scala's `Option`) which is reflected in my IDE's default warnings.

Passing nulls in place of a newly added optional parameters does not look good to me, hence plain `String` has been replaced with a dedicated type with an optional record type (which returns an `Optional<String>` and enforces its API users to handle the case with record type being absent. And that's the use case recommended by Java team members and authorities). 

Another motivation was to improve type safety, make method signatures more clear and avoid having some _util_ class with a static method that creates a subject from two strings.

mkubala commented on kcbq-confluent/src/main/java/com/wepay/kafka/connect/bigquery/schemaregistry/schemaretriever/SchemaRegistrySchemaRetriever.java:
```diff
@@ -23,7 +24,9 @@
 import java.io.IOException;
 
-import java.util.Map;
+import java.util.*;
```
Roger that 👍

mkubala commented on kcbq-confluent/src/main/java/com/wepay/kafka/connect/bigquery/schemaregistry/schemaretriever/SchemaRegistrySchemaRetriever.java:
```diff
@@ -79,9 +103,33 @@ public Schema retrieveSchema(TableId table, String topic, KafkaSchemaRecordType
+  private Map<String, TopicAndRecordName> getSubjects(List<String> topics, Map<Pattern, String> recordAliases) {
+    Collection<String> subjects;
+    try {
+      subjects = schemaRegistryClient.getAllSubjects();
```
Maybe there is a better solution for fetching all supported subjects but unfortunately I could not find it.

We are using Kafka topics to carry on some events, like user logged in, user paid, admin removed an account, etc. Each type of an event has its own record type and we rely on `TopicRecordNameStrategy` (`io.confluent.kafka.serializers.subject.TopicRecordNameStrategy`).
This might sound like a hack while it's one of the fully supported use cases (see https://www.confluent.io/blog/put-several-event-types-kafka-topic/).

Number of types of events (record types) might be humongous and increase in time (usually new feature introduces a couple of events). Maintaining a list of all supported record types for a topic in the connector's configuration would not only be cumbersome but also would require restarts and reconfiguration whenever you introduce new record type (which becomes more problematic when using some hosted kafka solutions).

That's why I decided to rely on regular expressions which let me declare one or a couple of patterns (instead of dozens or hundreds) to rule them all. Also since creating missing tables on the fly is supported, there is no need for restarts.

Unfortunately we cannot query schema reqistry using those patterns and we do not know exact record names in advance.
The only solution is to fetch them all and filter, using regular expressions defined in the config.

Alternatively we could just forget about creating tables at startup time and create them on the fly, when a new record type appears (which I've already implemented in this PR and can rebase on top of the recent changes introduced in #233 ) 🤔 
What do you think?

mkubala commented:
Thanks for you feedback!

I've addressed most of the comments. As the next step I'm going to rebase my work on top of @bingqinzhou's work and prepare a few, better grained and smaller PRs.
I'm not going to rebase this PR, since force push will discard all the valuable comments.

I hope that my answers in this PR made my motivation more clear and you will notice the dependencies between all the changes. 
Beside support for handling multiple record types, being able to create tables on the fly and handle nested empty structures in both record and schema converters is essential to make the whole feature reasonably usable on production. 
So you might expect at least 3 smaller PRs from me and it would be much appreciated if they could be released together.
